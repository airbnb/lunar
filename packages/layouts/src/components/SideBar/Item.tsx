import React from 'react';
import withStyles, { WithStylesProps } from '@airbnb/lunar/lib/composers/withStyles';
import ButtonOrLink, { ButtonOrLinkTypes } from '@airbnb/lunar/lib/components/private/ButtonOrLink';
import iconComponent from '@airbnb/lunar/lib/prop-types/iconComponent';
import Text from '@airbnb/lunar/lib/components/Text';

export type Props = {
  /** Mark the item as active. */
  active?: boolean;
  /** Render as an anchor link with a URL. */
  href?: string;
  /** Icon to display above the label. */
  icon: NonNullable<React.ReactElement<any>>;
  /** Item label. */
  label?: React.ReactNode;
  /** Callback fired when the element is clicked. */
  onClick?: (event: React.MouseEvent<ButtonOrLinkTypes>) => void;
};

/** A clickable item within the sidebar navigation menu. */
export class SideBarItem extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    icon: iconComponent.isRequired,
  };

  static defaultProps = {
    active: false,
    label: null,
  };

  render() {
    const { cx, active, label, href, icon, onClick, styles } = this.props;

    return (
      <li role="none">
        <ButtonOrLink
          role="menuitem"
          href={href}
          onClick={onClick}
          className={cx(styles.item, active && styles.item_active)}
        >
          <span className={cx(styles.icon)}>{React.cloneElement(icon, { size: '2em' })}</span>

          {label && (
            <span className={cx(styles.label)}>
              <Text micro bold uppercased inverted>
                {label}
              </Text>
            </span>
          )}
        </ButtonOrLink>
      </li>
    );
  }
}

export default withStyles(({ unit, color, pattern, transition }) => ({
  item: {
    ...pattern.resetButton,
    ...transition.box,
    paddingTop: unit * 1.5,
    paddingBottom: unit * 1.5,
    border: 0,
    display: 'block',
    textAlign: 'center',
    width: '100%',
    color: color.accent.bg,
    background: color.core.neutral[5],

    ':hover': {
      color: color.accent.bgHover,
      background: color.core.neutral[6],
    },
  },

  item_active: {
    background: color.core.neutral[6],
  },

  icon: {
    display: 'block',

    '@selectors': {
      '> svg': {
        display: 'inline-block',
        margin: 'auto',
      },
    },
  },

  label: {
    display: 'block',
    paddingTop: unit * 0.25,
  },
}))(SideBarItem);
