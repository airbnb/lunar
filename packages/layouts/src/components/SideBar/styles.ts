import { StyleSheet } from '@airbnb/lunar/lib/hooks/useStyles';

export const styleSheetSideBar: StyleSheet = ({ unit, color }) => ({
  bar: {
    width: unit * 8,
    height: '100%',
    background: color.core.neutral[5],
  },

  list: {
    display: 'flex',
    flexDirection: 'column',
    margin: 0,
    padding: 0,

    '@selectors': {
      '> li': {
        listStyle: 'none',
        margin: 0,
      },
    },
  },
});

export const styleSheetItem: StyleSheet = ({ unit, color, pattern, transition }) => ({
  item: {
    ...pattern.resetButton,
    ...transition.box,
    paddingTop: unit * 1.5,
    paddingBottom: unit * 1.5,
    border: 0,
    display: 'block',
    textAlign: 'center',
    width: '100%',
    color: color.accent.bg,
    background: color.core.neutral[5],

    ':hover': {
      color: color.accent.bgHover,
      background: color.core.neutral[6],
    },
  },

  item_active: {
    background: color.core.neutral[6],
  },

  icon: {
    display: 'block',

    '@selectors': {
      '> svg': {
        display: 'inline-block',
        margin: 'auto',
      },
    },
  },

  label: {
    display: 'block',
    paddingTop: unit * 0.25,
  },
});
