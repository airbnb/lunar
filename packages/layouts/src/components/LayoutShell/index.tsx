import React from 'react';
import { elementType } from 'airbnb-prop-types';
import useStyles, { Theme } from '@airbnb/lunar/lib/hooks/useStyles';
import SideBar from '../SideBar';

const styleSheet = ({ color }: Theme) => ({
  shell: {
    display: 'flex',
    width: '100%',
    minHeight: '100vh',
    background: color.accent.bg,
    justifyContent: 'space-between',
  },

  sideBar: {
    flexGrow: 0,
    flexShrink: 0,
    width: 'auto',
  },
});

export type Props = {
  /** The page content. */
  children: NonNullable<React.ReactNode>;
  /** Navigation side bar to display before the content. */
  sideBar?: React.ReactNode;
  /** Navigation top bar to display above the content. */
  // topBar?: React.ReactNode;
};

/** Layout shell that wraps an entire application, providing optional side and top nav bars. */
export default function LayoutShell({ children, sideBar }: Props) {
  const [styles, cx] = useStyles(styleSheet);

  return (
    <div className={cx(styles.shell)}>
      {sideBar && <aside className={cx(styles.sideBar)}>{sideBar}</aside>}

      <div>{children}</div>
    </div>
  );
}

LayoutShell.propTypes = {
  sideBar: elementType(SideBar),
};
