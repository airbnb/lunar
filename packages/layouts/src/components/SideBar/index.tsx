import React from 'react';
import withStyles, { WithStylesProps } from '@airbnb/lunar/lib/composers/withStyles';
import Item from './Item';

export type Props = {
  /** Accessibility label. */
  accessibilityLabel: string;
  /** Navigation items to render within the bar. */
  children: NonNullable<React.ReactNode>;
};

/** A vertical sidebar navigation menu. Primarily aligned on the left viewport. */
export class SideBar extends React.Component<Props & WithStylesProps> {
  render() {
    const { cx, accessibilityLabel, children, styles } = this.props;

    return (
      <nav className={cx(styles.bar)}>
        <ul role="menubar" aria-label={accessibilityLabel} className={cx(styles.list)}>
          {children}
        </ul>
      </nav>
    );
  }
}

export { Item };

export default withStyles(({ unit, color }) => ({
  bar: {
    width: unit * 8,
    height: '100%',
    background: color.core.neutral[5],
  },

  list: {
    display: 'flex',
    flexDirection: 'column',
    margin: 0,
    padding: 0,

    '@selectors': {
      '> li': {
        listStyle: 'none',
        margin: 0,
      },
    },
  },
}))(SideBar);
