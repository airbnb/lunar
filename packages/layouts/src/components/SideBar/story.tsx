import React from 'react';
import IconUser from '@airbnb/lunar-icons/lib/general/IconUser';
import IconSettings from '@airbnb/lunar-icons/lib/interface/IconSettings';
import IconChat from '@airbnb/lunar-icons/lib/interface/IconChat';
import SideBar, { Item } from '.';

export default {
  title: 'Layouts/SideBar',
  parameters: {
    inspectComponents: [SideBar, Item],
  },
};

export function aVerticalSidebarWithMenuItems() {
  return (
    <SideBar accessibilityLabel="Side menu">
      <Item icon={<IconUser accessibilityLabel="Profile" />} />
      <Item icon={<IconSettings accessibilityLabel="Settings" />} />
      <Item icon={<IconChat accessibilityLabel="Chat" />} />
    </SideBar>
  );
}

aVerticalSidebarWithMenuItems.story = {
  name: 'A vertical sidebar with menu items.',
};

export function withLabelsAndAnActiveState() {
  return (
    <SideBar accessibilityLabel="Side menu">
      <Item icon={<IconUser accessibilityLabel="Profile" />} label="Stats" />
      <Item active icon={<IconSettings accessibilityLabel="Settings" />} label="Config" />
      <Item icon={<IconChat accessibilityLabel="Chat" />} label="Chat" />
    </SideBar>
  );
}

withLabelsAndAnActiveState.story = {
  name: 'With labels and an active state.',
};
