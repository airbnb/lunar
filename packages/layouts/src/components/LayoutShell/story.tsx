import React from 'react';
import IconSettings from '@airbnb/lunar-icons/lib/interface/IconSettings';
import LoremIpsum from ':storybook/components/LoremIpsum';
import SideBar, { Item } from '../SideBar';
import LayoutShell from '.';

export default {
  title: 'Layouts/LayoutShell',
  parameters: {
    inspectComponents: [LayoutShell],
  },
};

export function standardShellWithNoTopOrSideBars() {
  return (
    <LayoutShell>
      <LoremIpsum />
      <LoremIpsum />
      <LoremIpsum />
      <LoremIpsum />
    </LayoutShell>
  );
}

standardShellWithNoTopOrSideBars.story = {
  name: 'Standard shell with no top or side bars.',
};

export function withSideBarAndNoBackgroundColor() {
  return (
    <LayoutShell
      sideBar={
        <SideBar accessibilityLabel="Nav">
          <Item icon={<IconSettings accessibilityLabel="Settings" />} />
        </SideBar>
      }
    >
      <LoremIpsum />
      <LoremIpsum />
      <LoremIpsum />
      <LoremIpsum />
    </LayoutShell>
  );
}

withSideBarAndNoBackgroundColor.story = {
  name: 'With side bar and no background color.',
};
