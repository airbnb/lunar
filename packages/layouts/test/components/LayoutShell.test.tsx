import React from 'react';
import { shallow } from 'enzyme';
import SideBar from '../../src/components/SideBar';
import LayoutShell from '../../src/components/LayoutShell';

describe('<LayoutShell />', () => {
  it('errors if invalid `sideBar` is passed', () => {
    expect(() => {
      shallow(<LayoutShell sideBar="Sidebar">Child</LayoutShell>);
    }).toThrowError();
  });

  it('renders with content', () => {
    const wrapper = shallow(<LayoutShell>Child</LayoutShell>);

    expect(
      wrapper
        .find('div')
        .at(1)
        .text(),
    ).toBe('Child');
  });

  it('renders a sidebar', () => {
    const sidebar = <SideBar accessibilityLabel="Test">Item</SideBar>;
    const wrapper = shallow(<LayoutShell sideBar={sidebar}>Child</LayoutShell>);

    expect(wrapper.find('aside')).toHaveLength(1);
    expect(wrapper.find('aside').contains(sidebar)).toBe(true);
  });
});
