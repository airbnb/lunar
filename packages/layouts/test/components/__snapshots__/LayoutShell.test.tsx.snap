// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayoutShell /> errors if invalid \`sideBar\` is passed 1`] = `
"Warning: Failed prop type: Invalid prop \`sideBar\` of type \`string\` supplied to \`LayoutShell\`, expected a single ReactElement.
    in LayoutShell"
`;
